File: app/src/main/java/info/nukoneko/cuc/android/kidspos/ui/main/calculate/CalculatorDialogFragment.java
package info.nukoneko.cuc.android.kidspos.ui.main.calculate;

import android.app.Dialog;
import android.app.ProgressDialog;
import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.os.Parcelable;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.Window;
import android.widget.Toast;

import java.util.List;

import info.nukoneko.cuc.android.kidspos.KidsPOSApplication;
import info.nukoneko.cuc.android.kidspos.R;
import info.nukoneko.cuc.android.kidspos.databinding.FragmentCalculatorBinding;
import info.nukoneko.cuc.android.kidspos.entity.Item;
import info.nukoneko.cuc.android.kidspos.entity.Sale;
import info.nukoneko.cuc.android.kidspos.entity.Staff;
import info.nukoneko.cuc.android.kidspos.entity.Store;
import info.nukoneko.cuc.android.kidspos.event.SuccessSentSaleEvent;
import info.nukoneko.cuc.android.kidspos.ui.common.BaseDialogFragment;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.functions.Consumer;

public final class CalculatorDialogFragment extends BaseDialogFragment implements CalculatorLayout.Listener, AccountResultDialogFragment.Listener {
    private static final String EXTRA_SUM_RIVER = "sum_price";
    private static final String EXTRA_SALE_ITEMS = "sales_model";

    private int mReceiveMoney = 0;
    private FragmentCalculatorBinding mBinding;
    private AccountResultDialogFragment mDialogFragment = null;

    @NonNull
    public static CalculatorDialogFragment newInstance(final int sumRiver, @NonNull final List<Item> saleItems) {
        final Bundle bundle = new Bundle();
        bundle.putInt(EXTRA_SUM_RIVER, sumRiver);
        bundle.putParcelableArray(EXTRA_SALE_ITEMS, saleItems.toArray(new Item[saleItems.size()]));
        final CalculatorDialogFragment fragment = new CalculatorDialogFragment();
        fragment.setArguments(bundle);
        return fragment;
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        mBinding = DataBindingUtil.inflate(LayoutInflater.from(getContext()), R.layout.fragment_calculator, container, true);
        mBinding.calculatorLayout.setOnCalculatorClickListener(this);
        return mBinding.getRoot();
    }

    @Override
    public void onActivityCreated(Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);

        getDialog().setCancelable(false);

        final Window dialogWindow = getDialog().getWindow();
        if (dialogWindow != null) {
            dialogWindow.setLayout(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT);
        }

        mBinding.sumRiver.setText(String.valueOf(getSumRiver()));
        mBinding.done.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (!isValid()) return;

                mDialogFragment = AccountResultDialogFragment.newInstance(getSumRiver(), mReceiveMoney);
                mDialogFragment.setCancelable(false);
                mDialogFragment.setListener(CalculatorDialogFragment.this);
                mDialogFragment.show(getChildFragmentManager(), "yesNoDialog");
            }
        });
        mBinding.back.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                dismiss();
            }
        });
    }

    @Override
    public void onClickNumber(int number) {
        if (mReceiveMoney > 100000) return;

        if (mReceiveMoney == 0) {
            mReceiveMoney = number;
        } else {
            mReceiveMoney = mReceiveMoney * 10 + number;
        }
        mBinding.receiveRiver.setText(String.valueOf(mReceiveMoney));
    }

    @Override
    public void onClickClear(View view) {
        if (10 > mReceiveMoney) {
            mReceiveMoney = 0;
        } else {
            mReceiveMoney = (int) Math.floor(mReceiveMoney / 10);
        }
        mBinding.receiveRiver.setText(String.valueOf(mReceiveMoney));
    }

    @Override
    public void onClickPositiveButton(Dialog dialog) {
        sendToServer();
    }

    @Override
    public void onClickNegativeButton(Dialog dialog) {
        dialog.cancel();
    }

    private int getSumRiver() {
        return getArguments().getInt(EXTRA_SUM_RIVER);
    }

    @SuppressWarnings("ConstantConditions")
    @NonNull
    private Item[] getSaleItems() {
        final Parcelable[] parcelables = getArguments().getParcelableArray(EXTRA_SALE_ITEMS);
        Item[] ret = new Item[parcelables.length];
        for (int i = 0; i < parcelables.length; i++) {
            ret[i] = (Item) parcelables[i];
        }
        return ret;
    }

    private boolean isValid() {
        return getSumRiver() <= this.mReceiveMoney;
    }

    private void finishFragment() {
        if (getContext() == null) return;
        KidsPOSApplication app = KidsPOSApplication.get(getContext());
        if (app == null) return;

        if (mDialogFragment != null) mDialogFragment.getDialog().cancel();
        app.postEvent(new SuccessSentSaleEvent());
        dismiss();
    }

    private void sendToServer() {
        if (getContext() == null) return;
        KidsPOSApplication app = KidsPOSApplication.get(getContext());
        if (app == null) return;

        StringBuilder sum = new StringBuilder();
        for (final Item item : getSaleItems()) {
            sum.append(String.valueOf(item.getId())).append(",");
        }
        sum = new StringBuilder(sum.substring(0, sum.length() - 1));
        final Staff staff = app.getCurrentStaff();
        final Store store = app.getCurrentStore();
        final String staffBarcode = staff == null ? "" : staff.getBarcode();
        final int storeId = store == null ? 0 : store.getId();

        if (app.isPracticeModeEnabled()) {
            Toast.makeText(getContext(), "練習モードのためレシートは出ません", Toast.LENGTH_SHORT).show();
            finishFragment();
        } else {
            final ProgressDialog progressDialog = new ProgressDialog(getContext());
            progressDialog.setTitle("送信しています");
            progressDialog.show();

            app.getSaleManager().createSale(mReceiveMoney, getSaleItems().length, getSumRiver(), sum.toString(), storeId, staffBarcode)
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribeOn(app.defaultSubscribeScheduler())
                    .subscribe(new Consumer<Sale>() {
                        @Override
                        public void accept(Sale sale) throws Exception {
                            progressDialog.dismiss();
                            finishFragment();
                        }
                    }, new Consumer<Throwable>() {
                        @Override
                        public void accept(Throwable throwable) throws Exception {
                            throwable.printStackTrace();
                            progressDialog.dismiss();
                            finishFragment();
                        }
                    });
        }
    }
}

File: app/src/main/java/info/nukoneko/cuc/android/kidspos/ui/main/calculate/CalculatorDialogFragment.kt
package info.nukoneko.cuc.android.kidspos.ui.main.calculate

import android.app.Dialog
import android.app.ProgressDialog
import android.arch.lifecycle.ViewModelProviders
import android.databinding.DataBindingUtil
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import info.nukoneko.cuc.android.kidspos.KidsPOSApplication
import info.nukoneko.cuc.android.kidspos.R
import info.nukoneko.cuc.android.kidspos.databinding.FragmentCalculatorBinding
import info.nukoneko.cuc.android.kidspos.entity.Item
import info.nukoneko.cuc.android.kidspos.event.SystemEvent
import info.nukoneko.cuc.android.kidspos.extensions.lazyWithArgs
import info.nukoneko.cuc.android.kidspos.ui.common.BaseDialogFragment
import io.reactivex.android.schedulers.AndroidSchedulers

class CalculatorDialogFragment : BaseDialogFragment(), AccountResultDialogFragment.Listener {

    private var deposit = 0
    private lateinit var binding: FragmentCalculatorBinding

    // TODO: coroutine + channel へ乗り換え
    private var resultDialogFragment: AccountResultDialogFragment? = null

    private val totalPrice: Int by lazyWithArgs(EXTRA_SUM_RIVER)
    private val items: List<Item> by lazyWithArgs(EXTRA_SALE_ITEMS)

    private val viewModel: CalculatorDialogViewModel by lazy {
        ViewModelProviders.of(this)[CalculatorDialogViewModel::class.java]
    }

    private val isValid: Boolean
        get() = totalPrice <= this.deposit

    private val calculatorListener = object: CalculatorLayout.Listener {
        override fun onNumberClick(number: Int) {
            if (deposit > 100000) return

            if (deposit == 0) {
                deposit = number
            } else {
                deposit = deposit * 10 + number
            }
            binding.receiveRiver.setText(deposit.toString())
        }

        override fun onClear(view: View) {
            if (10 > deposit) {
                deposit = 0
            } else {
                deposit = Math.floor((deposit / 10).toDouble()).toInt()
            }
            binding.receiveRiver.setText(deposit.toString())
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        binding = DataBindingUtil.inflate(LayoutInflater.from(context), R.layout.fragment_calculator, container, true)
        binding.calculatorLayout.listener = calculatorListener
        binding.viewModel = viewModel
        return binding.root
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)

        dialog.setCancelable(false)

        val dialogWindow = dialog.window
        dialogWindow?.setLayout(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT)

        binding.sumRiver.text = totalPrice.toString()
        binding.done.setOnClickListener(View.OnClickListener {
            if (!isValid) return@OnClickListener

            resultDialogFragment = AccountResultDialogFragment.newInstance(totalPrice, deposit)
            resultDialogFragment!!.isCancelable = false
            resultDialogFragment!!.setListener(this@CalculatorDialogFragment)
            resultDialogFragment!!.show(childFragmentManager, "yesNoDialog")
        })
        binding.back.setOnClickListener { dismiss() }
    }

    override fun onClickPositiveButton(dialog: Dialog) {
        sendToServer()
    }

    override fun onClickNegativeButton(dialog: Dialog) {
        dialog.cancel()
    }

    private fun finishFragment() {
        val app = KidsPOSApplication[context] ?: return
        resultDialogFragment?.dialog?.cancel()
        app.postEvent(SystemEvent.SentSaleSuccess)
        dismiss()
    }

    private fun sendToServer() {
        val app = KidsPOSApplication[context] ?: return

        var sum = StringBuilder()
        for ((id) in items) {
            sum.append(id.toString()).append(",")
        }
        sum = StringBuilder(sum.substring(0, sum.length - 1))
        val staff = app.storeManager.lastStaff
        val store = app.storeManager.lastStore
        val staffBarcode = staff?.barcode ?: ""
        val storeId = store?.id ?: 0

        if (app.isPracticeModeEnabled) {
            Toast.makeText(context, "練習モードのためレシートは出ません", Toast.LENGTH_SHORT).show()
            finishFragment()
        } else {
            val progressDialog = ProgressDialog(context)
            progressDialog.setTitle("送信しています")
            progressDialog.show()

            app.saleManager.createSale(deposit, items.size, totalPrice, sum.toString(), storeId, staffBarcode)
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribeOn(app.getDefaultSubscribeScheduler())
                    .subscribe({ (id, barcode, createdAt, points, price, items, storeId1, staffId) ->
                        progressDialog.dismiss()
                        finishFragment()
                    }, { throwable ->
                        throwable.printStackTrace()
                        progressDialog.dismiss()
                        finishFragment()
                    })
        }
    }

    companion object {
        private const val EXTRA_SUM_RIVER = "sum_price"
        private const val EXTRA_SALE_ITEMS = "sales_model"

        fun newInstance(sumRiver: Int, saleItems: List<Item>) = CalculatorDialogFragment().apply {
            arguments = Bundle().apply {
                putInt(EXTRA_SUM_RIVER, sumRiver)
                putParcelableArray(EXTRA_SALE_ITEMS, saleItems.toTypedArray())
            }
        }
    }
}
